File: test_twelvedata_days.R
# Twelve Data API - 验证免费版能获取多少天的K线数据
library(httr)
library(jsonlite)

# 读取 .Renviron
readRenviron(".Renviron")

API_KEY <- Sys.getenv("TWELVEDATA_API_KEY")

if (API_KEY == "") {
  stop("TWELVEDATA_API_KEY not found!")
}

cat("API Key:", substr(API_KEY, 1, 15), "...\n\n")

BASE_URL <- "https://api.twelvedata.com/time_series"

# 测试1: 不设置日期范围，看看默认返回多少条
cat(paste(rep("=", 60), collapse = ""), "\n")
cat("测试1: 不设置日期范围 (默认 outputsize)\n")
cat(paste(rep("=", 60), collapse = ""), "\n")

params1 <- list(
  symbol = "AAPL",
  interval = "1day",
  apikey = API_KEY
)

resp1 <- GET(BASE_URL, query = params1)
data1 <- content(resp1, as = "parsed")

if (!is.null(data1$values)) {
  cat("✅ 返回条数:", length(data1$values), "条\n")
  cat("   日期范围:", data1$values[[length(data1$values)]]$datetime,
      "至", data1$values[[1]]$datetime, "\n\n")
} else {
  cat("❌ 错误:", toJSON(data1), "\n\n")
}

# 测试2: 设置大日期范围，获取尽可能多的数据
cat(paste(rep("=", 60), collapse = ""), "\n")
cat("测试2: 设置大日期范围 (2021-01-01 至 2026-03-13)\n")
cat(paste(rep("=", 60), collapse = ""), "\n")

params2 <- list(
  symbol = "AAPL",
  interval = "1day",
  start_date = "2021-01-01",
  end_date = "2026-03-13",
  apikey = API_KEY
)

resp2 <- GET(BASE_URL, query = params2)
data2 <- content(resp2, as = "parsed")

if (!is.null(data2$values)) {
  cat("✅ 返回条数:", length(data2$values), "条\n")
  cat("   日期范围:", data2$values[[length(data2$values)]]$datetime,
      "至", data2$values[[1]]$datetime, "\n")

  # 计算大约多少年的数据
  first_date <- as.Date(data2$values[[1]]$datetime)
  last_date <- as.Date(data2$values[[length(data2$values)]]$datetime)
  days <- as.numeric(first_date - last_date)
  years <- days / 365
  cat("   约", round(years, 1), "年的数据\n\n")
} else {
  cat("❌ 错误:", toJSON(data2), "\n\n")
}

# 测试3: 不限制outputsize最大值
cat(paste(rep("=", 60), collapse = ""), "\n")
cat("测试3: 请求更大的日期范围\n")
cat(paste(rep("=", 60), collapse = ""), "\n")

params3 <- list(
  symbol = "AAPL",
  interval = "1day",
  start_date = "2020-01-01",
  end_date = "2026-03-13",
  apikey = API_KEY
)

resp3 <- GET(BASE_URL, query = params3)
data3 <- content(resp3, as = "parsed")

if (!is.null(data3$values)) {
  cat("✅ 返回条数:", length(data3$values), "条\n")
  cat("   日期范围:", data3$values[[length(data3$values)]]$datetime,
      "至", data3$values[[1]]$datetime, "\n")

  first_date <- as.Date(data3$values[[1]]$datetime)
  last_date <- as.Date(data3$values[[length(data3$values)]]$datetime)
  days <- as.numeric(first_date - last_date)
  years <- days / 365
  cat("   约", round(years, 1), "年的数据\n\n")
} else {
  cat("❌ 错误:", toJSON(data3), "\n\n")
}

cat("测试完成!\n")

File: test_twelvedata_days.py
# Twelve Data API - 验证免费版能获取多少天的K线数据
import os
import requests

# 加载 .Renviron
def load_renviron(env_file='.Renviron'):
    """读取 .Renviron 文件"""
    try:
        with open(env_file, 'r') as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith('#'):
                    if '=' in line:
                        k, v = line.split('=', 1)
                        os.environ[k.strip()] = v.strip().strip('"')
    except FileNotFoundError:
        pass

load_renviron()

API_KEY = os.getenv("TWELVEDATA_API_KEY")

if not API_KEY:
    print("❌ TWELVEDATA_API_KEY not found!")
    exit(1)

print(f"API Key: {API_KEY[:10]}...")
print()

BASE_URL = "https://api.twelvedata.com/time_series"

# 测试1: 不设置日期范围，看看默认返回多少条
print("=" * 60)
print("测试1: 不设置日期范围 (默认 outputsize)")
print("=" * 60)
params = {
    "symbol": "AAPL",
    "interval": "1day",
    "apikey": API_KEY
}
resp = requests.get(BASE_URL, params=params)
data = resp.json()

if "values" in data:
    print(f"✅ 返回条数: {len(data['values'])} 条")
    print(f"   日期范围: {data['values'][-1]['datetime']} 至 {data['values'][0]['datetime']}")
else:
    print(f"❌ 错误: {data}")

print()

# 测试2: 设置大日期范围，获取尽可能多的数据
print("=" * 60)
print("测试2: 设置大日期范围 (2021-01-01 至 2026-03-13)")
print("=" * 60)
params2 = {
    "symbol": "AAPL",
    "interval": "1day",
    "start_date": "2021-01-01",
    "end_date": "2026-03-13",
    "apikey": API_KEY
}
resp2 = requests.get(BASE_URL, params=params2)
data2 = resp2.json()

if "values" in data2:
    print(f"✅ 返回条数: {len(data2['values'])} 条")
    print(f"   日期范围: {data2['values'][-1]['datetime']} 至 {data2['values'][0]['datetime']}")

    # 计算大约多少年的数据
    from datetime import datetime
    first_date = datetime.strptime(data2['values'][0]['datetime'], "%Y-%m-%d")
    last_date = datetime.strptime(data2['values'][-1]['datetime'], "%Y-%m-%d")
    days = (first_date - last_date).days
    years = days / 365
    print(f"   约 {years:.1f} 年的数据")
else:
    print(f"❌ 错误: {data2}")

print()

# 测试3: 使用 outputsize 参数
print("=" * 60)
print("测试3: outputsize=5000 (最大可能值)")
print("=" * 60)
params3 = {
    "symbol": "AAPL",
    "interval": "1day",
    "outputsize": 5000,
    "apikey": API_KEY
}
resp3 = requests.get(BASE_URL, params=params3)
data3 = resp3.json()

if "values" in data3:
    print(f"✅ 返回条数: {len(data3['values'])} 条")
    print(f"   日期范围: {data3['values'][-1]['datetime']} 至 {data3['values'][0]['datetime']}")
else:
    print(f"❌ 错误: {data3}")